File: frameworks/Java/vertx/src/main/java/vertx/rocker/BufferRockerOutput.java
package vertx.rocker;

import com.fizzed.rocker.ContentType;
import com.fizzed.rocker.RockerOutput;
import com.fizzed.rocker.RockerOutputFactory;
import io.netty.buffer.ByteBuf;
import io.netty.buffer.Unpooled;
import io.vertx.core.buffer.Buffer;
import io.vertx.core.net.impl.PartialPooledByteBufAllocator;

import java.io.IOException;
import java.nio.charset.Charset;
import java.nio.charset.StandardCharsets;

public class BufferRockerOutput implements RockerOutput<BufferRockerOutput> {

  public static RockerOutputFactory<BufferRockerOutput> factory(ContentType contentType) {
    BufferRockerOutput output = new BufferRockerOutput(contentType);
    return (_contentType, charsetName) -> {
      output.reset();
      return output;
    };
  }

  private final ByteBuf buff = PartialPooledByteBufAllocator.UNPOOLED.directBuffer();
  private final Buffer buffer = Buffer.buffer(buff);
  private final ContentType contentType;

  BufferRockerOutput(ContentType contentType) {
    this.contentType = contentType;
  }

  private void reset() {
    buff.resetReaderIndex();
    buff.resetWriterIndex();
  }

  @Override
  public BufferRockerOutput w(byte[] bytes) throws IOException {
    buffer.appendBytes(bytes);
    return this;
  }

  @Override
  public BufferRockerOutput w(String s) throws IOException {
    buffer.appendString(s);
    return this;
  }

  @Override
  public ContentType getContentType() {
    return contentType;
  }

  @Override
  public Charset getCharset() {
    return StandardCharsets.UTF_8;
  }

  @Override
  public int getByteLength() {
    return buffer.length();
  }

  public Buffer buffer() {
    return buffer;
  }
}

File: frameworks/Kotlin/vertx-web-kotlin-dsljson/src/main/kotlin/com/example/starter/helpers/BufferRockerOutput.kt
package com.example.starter.helpers

import com.fizzed.rocker.ContentType
import com.fizzed.rocker.RockerOutput
import com.fizzed.rocker.RockerOutputFactory
import io.netty.buffer.ByteBuf
import io.vertx.core.buffer.Buffer
import io.vertx.core.impl.buffer.VertxByteBufAllocator
import io.vertx.core.internal.buffer.BufferInternal
import java.nio.charset.Charset
import java.nio.charset.StandardCharsets

class BufferRockerOutput private constructor(
    private val contentType: ContentType
) : RockerOutput<BufferRockerOutput> {

    private val byteBuf: ByteBuf = VertxByteBufAllocator.DEFAULT.directBuffer()
    private val buffer: Buffer = BufferInternal.buffer(byteBuf)

    private fun reset() {
        byteBuf.resetReaderIndex()
        byteBuf.resetWriterIndex()
    }

    override fun w(bytes: ByteArray): BufferRockerOutput {
        buffer.appendBytes(bytes)
        return this
    }

    override fun w(s: String): BufferRockerOutput {
        buffer.appendString(s)
        return this
    }

    override fun getContentType(): ContentType = contentType

    override fun getCharset(): Charset = StandardCharsets.UTF_8

    override fun getByteLength(): Int = buffer.length()

    fun buffer(): Buffer = buffer

    companion object {
        fun factory(contentType: ContentType): RockerOutputFactory<BufferRockerOutput> {
            val output = BufferRockerOutput(contentType)
            return RockerOutputFactory { _, _ ->
                output.reset()
                output
            }
        }
    }
}